### PR TITLE
Adds support for multi tables

### PR DESCRIPTION
**Pull request by mathetake**
This commit adds support for multiple tables per module.
Notably, if the `WithFeatureReferenceTypes` is enabled,
`call_indirect`, `table.init` and `table.copy` instructions 
can reference non-zero indexed tables.

part of #484

### CODE REVIEW

**Comment by mathetake** [comment]
so the interpreter done... will start on JIT in a few hours

File: internal/integration_test/spectest/spectest.go [hunk 1]
```diff
@@ -308,6 +308,9 @@ func Run(t *testing.T, testDataFS embed.FS, newEngine func(wasm.Features) wasm.E
 		require.NoError(t, json.Unmarshal(raw, &base))
 
 		wastName := basename(base.SourceFile)
+		if wastName != "table_init.wast" {
```
Comment (mathetake) [inline internal/integration_test/spectest/spectest.go:311]: oh man

File: config.go [hunk 2]
```diff
@@ -27,7 +27,7 @@ type RuntimeConfig interface {
 	//
 	// Here are the notable effects:
 	// * Adds `memory.fill`, `memory.init`, `memory.copy` and `data.drop` instructions.
-	// * Adds `table.fill`, `table.init`, `table.copy` and `elem.drop` instructions.
```
Comment (mathetake) [inline config.go:30]: table.init is actually depending on the reference type (as it takes reference typed value as an argument) so essentially in ReferenceType proposal

**Comment by mathetake** [comment]
ok now backfilling more cases..

File: config.go (hunk 2 above)
Comment (codefromthecrypt) [reply config.go:30]: maybe the feature should say it is dependent on another (ex enabling this enables the other, and visa versa, disabling retypes disables this as it is not functional without it)

**Comment by mathetake** [comment]
ok ready!

File: config.go [hunk 3]
```diff
@@ -71,6 +72,31 @@ type RuntimeConfig interface {
...
+	//   * `table.get`
+	//   * `table.grow`
+	//   * `table.set`
+	//   * `table.size`
+	// * Support for multiple tables per module:
+	//   * `call_indirect`, `table.init`, `table.copy` and `elem.drop` instructions can take non-zero table index.
+	//   * Element segments can take non-zero table index.
+	//
+	//
+	// Note: "reference-types" is mixed with the "bulk-memory-operations" proposal
```
Comment (codefromthecrypt) [inline config.go:92]: oops ;)

File: internal/wasm/binary/section.go [hunk 4]
```diff
@@ -240,8 +250,11 @@ func encodeCodeSection(code []*wasm.Code) []byte {
 //
 // See encodeTable
 // See https://www.w3.org/TR/2019/REC-wasm-core-1-20191205/#table-section%E2%91%A0
-func encodeTableSection(table *wasm.Table) []byte {
-	contents := append([]byte{1}, encodeTable(table)...)
+func encodeTableSection(tables []*wasm.Table) []byte {
+	var contents []byte
+	for _, table := range tables {
+		contents = append([]byte{1}, encodeTable(table)...)
```
Comment (codefromthecrypt) [inline internal/wasm/binary/section.go:256]: comment why 1

File: internal/wasm/func_validation_test.go [hunk 5]
```diff
@@ -486,119 +486,153 @@ func TestModule_ValidateFunction_BulkMemoryOperations(t *testing.T) {
...
+			},
+			{
+				body:        []byte{OpcodeMiscPrefix, OpcodeMiscTableCopy, 3},
+				flag:        FeatureBulkMemoryOperations | FeatureReferenceTypes,
+				tables:      []*Table{{}, {}, {}, {}},
+				expectedErr: "failed to read source table index for table.copy: EOF",
+			},
+			{
+				body:        []byte{OpcodeMiscPrefix, OpcodeMiscTableCopy, 0, 3},
+				flag:        FeatureBulkMemoryOperations,
```
Comment (codefromthecrypt) [inline internal/wasm/func_validation_test.go:610]:
```suggestion
				flag:        FeatureBulkMemoryOperations, // Multiple tables require FeatureReferenceTypes.
```

File: internal/wasm/table.go [hunk 6]
```diff
@@ -41,9 +57,15 @@ const (
 type ElementSegment struct {
 	// OffsetExpr returns the table element offset to apply to Init indices.
 	// Note: This can be validated prior to instantiation unless it includes OpcodeGlobalGet (an imported global).
-	// Note: This is nil if the Mode is either passive or declarative.
+	// Note: This is not nil if and only if the Mode is active.
```
Comment (codefromthecrypt) [inline internal/wasm/table.go:60]:
```suggestion
	// Note: This is only set when Mode is active.
```

**Review by codefromthecrypt — APPROVED** [review]
Looks good. Maybe add one toehold wazeroir compiler test that uses externref and/or multiple table (ex maybe call_indirect)

**Pull request merged**